libcontainer/state_linux.go

```diff
@@ -0,0 +1,216 @@
+// +build linux
+
+package libcontainer
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/opencontainers/runc/libcontainer/configs"
+)
+
+func newStateError(from, to containerState) error {
+	return &stateError{
+		From: from.status().String(),
+		To:   to.status().String(),
+	}
+}
+
+// stateError is returned when an invalid state transition happens from one
+// state to another.
+type stateError struct {
+	From string
+	To   string
+}
+
+func (s *stateError) Error() string {
+	return fmt.Sprintf("invalid state transition from %s to %s", s.From, s.To)
+}
+
+type containerState interface {
+	transition(containerState) error
+	destroy() error
+	status() Status
+}
+
+// destroyedState represents a container is a stopped/destroyed state.
+type destroyedState struct {
+	c *linuxContainer
+}
+
+func (b *destroyedState) status() Status {
+	return Destroyed
+}
+
+func (b *destroyedState) transition(s containerState) error {
+	switch s.(type) {
+	case *runningState:
+		b.c.state = s
+		return nil
+	case *restoredState:
+		b.c.state = s
+		return nil
+	case *destroyedState:
+		return nil
+	}
+	return newStateError(b, s)
+}
+
+func (b *destroyedState) destroy() error {
+	if !b.c.config.Namespaces.Contains(configs.NEWPID) {
+		if err := killCgroupProcesses(b.c.cgroupManager); err != nil {
+			logrus.Warn(err)
+		}
+	}
+	err := b.c.cgroupManager.Destroy()
+	if rerr := os.RemoveAll(b.c.root); err == nil {
+		err = rerr
+	}
+	b.c.initProcess = nil
+	if herr := b.runHooks(); err == nil {
+		err = herr
+	}
+	return err
+}
+
+func (b *destroyedState) runHooks() error {
+	if b.c.config.Hooks != nil {
+		s := configs.HookState{
+			Version: b.c.config.Version,
+			ID:      b.c.id,
+			Root:    b.c.config.Rootfs,
+		}
+		for _, hook := range b.c.config.Hooks.Poststop {
+			if err := hook.Run(s); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// runningState represents a container that is currently running.
+type runningState struct {
+	base containerState
```

Comment:
Why do we need the base here? From the patch it looks like we are using it for calling destroy on destroyed state. Could we have a common destroy helper function and call it from destroy() of runningState/destroyState instead?